File: bonus/c/data_structures/binary_search_tree.c
/*
  Binary search tree data structure example
  Copyright 2017, Sjors van Gelderen
*/

#include <stdlib.h>
#include <stdio.h>

// Binary search tree data structure
typedef struct BST
{
    int key;
    struct BST* left;
    struct BST* right;
} BST;

// Function prototypes
void print_in_order_traversal(BST* _tree);
BST* bst_create();
void bst_insert(BST** _tree, int _key);
void bst_search(BST* _tree, int _key);
void terminate(BST** _tree);

// Prints in order traversal sequence of a tree
void print_in_order_traversal(BST* _tree)
{
    if(_tree != NULL)
    {
	print_in_order_traversal(_tree->left);
	printf("%d, ", _tree->key);
	print_in_order_traversal(_tree->right);
    }
}

// Allocates memory for the BST and returns a pointer to it
BST* bst_create(int _key)
{
    BST* tree = malloc(sizeof(BST));
    if(tree == NULL)
    {
	printf("Failed to allocate memory for the BST\n");
	return NULL;
    }
    
    tree->key = _key;
    tree->left = NULL;
    tree->right = NULL;
    return tree;
}

/*
  Inserts a node into the BST
  Complexity: O(n)
*/
void bst_insert(BST** _tree, int _key)
{
    if((*_tree) == NULL)
    {
	printf("Inserting node with key %d\n", _key);
	*_tree = bst_create(_key);
    }
    else
    {
	if(_key < (*_tree)->key)
	{
	    printf("Moving left from %d\n", (*_tree)->key);
	    bst_insert(&(*_tree)->left, _key);
	}
	else if(_key > (*_tree)->key)
	{
	    printf("Moving right from %d\n", (*_tree)->key);
	    bst_insert(&(*_tree)->right, _key);
	}
	else
	{
	    printf("Duplicate entry, ignoring insert!\n");
	}
    }
}

/*
  Find a node in the BST
  Complexity: O(n)
*/
void bst_search(BST* _tree, int _key)
{
    if(_tree == NULL)
    {
	printf("Couldn't find %d in the tree\n", _key);
    }
    else
    {
	if(_key < _tree->key)
	{
	    printf("Moving left from %d\n", _tree->key);
	    bst_search(_tree->left, _key);
	}
	else if(_key > _tree->key)
	{
	    printf("Moving right from %d\n", _tree->key);
	    bst_search(_tree->right, _key);
	}
	else
	{
	    printf("Found key %d in tree!\n", _key);
	}
    }
}

/*
  Frees and deallocates tree memory
  Complexity: O(n)
*/
void terminate(BST** _tree)
{
    if(*_tree != NULL)
    {
	BST* left = (*_tree)->left;
	BST* right = (*_tree)->right;
	free(*_tree);
	*_tree = NULL;
	terminate(&left);
	terminate(&right);
    }
}

// Main program logic
int main()
{
    printf("BST data structure example - ");
    printf("Copyright 2017, Sjors van Gelderen\n\n");
    
    // Empty tree
    BST* tree = NULL;

    // Insert nodes, pass pointer to pointer to tree
    bst_insert(&tree, 52);
    bst_insert(&tree, 3);
    bst_insert(&tree, 999);
    bst_insert(&tree, 33);

    // Print an in-order traversal
    printf("Traversal: ");
    print_in_order_traversal(tree);
    printf("\n");

    // Try to find these values in the tree
    bst_search(tree, 23);
    bst_search(tree, 55);
    bst_search(tree, 3);
    bst_search(tree, 999);

    // Free and deallocate memory
    terminate(&tree);
    
    return EXIT_SUCCESS;
}

File: bonus/fsharp/data_structures/binary_search_tree.fs
(*
Binary search tree example
Copyright 2017, Sjors van Gelderen
*)

// Binary search tree type containing a generic type for values
type BST<'a> =
    { Value : 'a
      Left  : BST<'a> Option
      Right : BST<'a> Option }

(*
In-order traversal
Complexity: O(n)
*)
let rec in_order_traversal tree =
    match tree with
    | Some tree ->
        (in_order_traversal tree.Left) @ (tree.Value :: (in_order_traversal tree.Right))
    | None ->
        []

(*
Insertion operation
Complexity: O(n)
*)
let rec insert value tree =
    match tree with
    | Some tree ->
        if value < tree.Value then
            // Recursion on the left subtree
            Some { tree with Left = insert value tree.Left }
        else
            // Recursion on the right subtree
            Some { tree with Right = insert value tree.Right }
    | None ->
        // Return a tree
        Some { Value = value
               Left  = None
               Right = None }

(*
Delete operation
Complexity: O(n log n)
*)
let rec delete value tree : BST<'a> Option =
    match tree with
    | Some tree ->
        if tree.Value = value then
            if tree.Left = None && tree.Right = None then
                None
            elif tree.Left <> None && tree.Right <> None then
                // Locate successor
                let predicate = (fun elem -> elem = tree.Value)
                let traversal = in_order_traversal (Some tree)
                match List.tryFind predicate traversal with
                | Some successor ->
                    // Recursion on the rest of the subtree
                    Some { tree with Value = successor
                                     Right = delete successor tree.Right }
                | None ->
                    printfn "Failed to locate successor of %A" tree.Value
                    None
            elif tree.Left <> None then
                // Replace tree with left subtree
                tree.Left
            elif tree.Right <> None then
                // Replace tree with right subtree
                tree.Right
            else
                // Delete this tree
                None
        else
            if value < tree.Value then
                // Recursion on the left subtree
                Some { tree with Left = delete value tree.Left }
            else
                // Recursion on the right subtree
                Some { tree with Right = delete value tree.Right }
    | None ->
        // Value not found
        printfn "Value %A not found in tree..." value
        None

(*
Search operation
Complexity: O(n)
*)
let rec search value tree =
    match tree with
    | Some tree ->
        if tree.Value = value then
            // Value found
            printfn "Found %A in tree!" value
            true
        else
            if value < tree.Value then
                // Recursion on the left subtree
                search value tree.Left
            else
                // Recursion on the right subtree
                search value tree.Right
    | None ->
        // Value not found
        printfn "Value %A not found in tree..." value
        false

// Main program logic
[<EntryPoint>]
let main args =
    let tree = Some { Value = 10
                      Left  = None
                      Right = None }

    let tree =
        tree
        |> insert 5
        |> insert 16
        |> insert 2
        |> insert 7
        |> delete 4 // Deliberate failure
        |> delete 2
        |> delete 5
    
    search 7 tree  |> ignore
    search 20 tree |> ignore // Deliberate failure
    search 16 tree |> ignore

    List.iter (fun elem -> printfn "%A" elem) <| in_order_traversal tree
    0
